File: core/internal/leet/runoverviewsidebarnav.go
package leet

const (
	// Each section's default share of the sidebar's vertical space when
	// the sections want more rows than fit. The shares sum to 1, so a
	// value reads directly as that section's slice of the area, in the
	// same unit as the dragged overview_* fractions.
	sectionWeightEnvironment = 0.20
	sectionWeightConfig      = 0.35
	sectionWeightSummary     = 0.45

	// Minimum section height when visible (title + 1 item).
	sectionMinHeight = 2
)

// sectionWeights returns each visible section's share of the section area:
// the user-dragged fraction where set, otherwise the built-in default.
// proportionalShares normalizes the weights, so dragged sections keep their
// exact shares while all of them are set and every section degrades
// proportionally when the data changes underneath.
func (s *RunOverviewSidebar) sectionWeights(needs []int) []float64 {
	defaults := []float64{
		sectionWeightEnvironment,
		sectionWeightConfig,
		sectionWeightSummary,
	}

	var o LayoutOverrides
	if s.overridesSource != nil {
		o = s.overridesSource()
	}
	fracs := o.overviewFractions()

	weights := make([]float64, len(needs))
	for i, need := range needs {
		if need <= 0 {
			continue
		}
		if fracs[i] > 0 {
			weights[i] = fracs[i]
		} else {
			weights[i] = defaults[i]
		}
	}
	return weights
}

// sectionNeeds returns the rows each section can usefully fill (title +
// items), 0 for empty sections.
func (s *RunOverviewSidebar) sectionNeeds() []int {
	needs := make([]int, len(s.sections))
	for i := range s.sections {
		if itemCount := len(s.sections[i].FilteredItems); itemCount > 0 {
			needs[i] = itemCount + 1 // Title line.
		}
	}
	return needs
}

// updateSectionHeights divides the sidebar rows below the header among the
// sections and re-derives every section's page size from its height.
func (s *RunOverviewSidebar) updateSectionHeights() {
	if s.height == 0 {
		return
	}

	needs := s.sectionNeeds()
	heights := flexSectionHeights(s.sectionsArea(needs), s.sectionWeights(needs), needs)
	for i := range s.sections {
		s.sections[i].Height = heights[i]
	}

	s.updateItemsPerPage()
}

// sectionsArea returns the rows available to sections: the sidebar height
// minus the header block and one spacing row between adjacent sections.
func (s *RunOverviewSidebar) sectionsArea(needs []int) int {
	visible := 0
	for _, need := range needs {
		if need > 0 {
			visible++
		}
	}
	return s.height - s.headerLineCount() - max(visible-1, 0)
}

// flexSectionHeights divides area rows among sections proportionally to
// weights. Sections with need 0 are hidden and get no rows. A visible
// section is capped at its need (title + items) so surplus rows go to
// sections that can still grow, and floored at sectionMinHeight so it stays
// usable when squeezed. When even the minimums do not fit, the total
// overflows area and the renderer crops it (tiny-terminal degradation).
func flexSectionHeights(area int, weights []float64, needs []int) []int {
	heights := make([]int, len(needs))
	active := make([]int, 0, len(needs))
	for i, need := range needs {
		if need > 0 {
			active = append(active, i)
		}
	}

	remaining := area
	for len(active) > 0 {
		shares := proportionalShares(active, weights, remaining)

		// Sections offered more rows than their items can fill keep only
		// their need; the surplus re-divides among the rest next pass.
		if next := settleSections(active, heights, &remaining, func(k, i int) (int, bool) {
			return needs[i], shares[k] >= float64(needs[i])
		}); len(next) < len(active) {
			active = next
			continue
		}

		// Sections offered fewer rows than the minimum take the minimum
		// anyway; the deficit comes out of the rest next pass.
		if next := settleSections(active, heights, &remaining, func(k, i int) (int, bool) {
			return min(sectionMinHeight, needs[i]), shares[k] < sectionMinHeight
		}); len(next) < len(active) {
			active = next
			continue
		}

		// Steady state: every share fits between its minimum and its need.
		roundShares(active, shares, heights, remaining)
		break
	}

	return heights
}

// proportionalShares splits remaining rows across the active sections in
// proportion to their weights (evenly, if the weights sum to zero).
func proportionalShares(active []int, weights []float64, remaining int) []float64 {
	var weightSum float64
	for _, i := range active {
		weightSum += weights[i]
	}

	shares := make([]float64, len(active))
	for k, i := range active {
		if weightSum > 0 {
			shares[k] = float64(remaining) * weights[i] / weightSum
		} else {
			shares[k] = float64(remaining) / float64(len(active))
		}
	}
	return shares
}

// settleSections assigns a final height to each active section that decide
// settles and returns the still-active rest.
func settleSections(
	active []int,
	heights []int,
	remaining *int,
	decide func(k, i int) (int, bool),
) []int {
	rest := make([]int, 0, len(active))
	for k, i := range active {
		if h, settled := decide(k, i); settled {
			heights[i] = h
			*remaining -= h
		} else {
			rest = append(rest, i)
		}
	}
	return rest
}

// roundShares converts fractional shares to whole rows, handing the leftover
// rows to the largest fractional remainders (ties to the earlier section).
func roundShares(active []int, shares []float64, heights []int, remaining int) {
	leftover := remaining
	fracs := make([]float64, len(active))
	for k, i := range active {
		heights[i] = int(shares[k])
		fracs[k] = shares[k] - float64(heights[i])
		leftover -= heights[i]
	}

	for ; leftover > 0; leftover-- {
		best := 0
		for k := range fracs {
			if fracs[k] > fracs[best] {
				best = k
			}
		}
		heights[active[best]]++
		fracs[best] = -1
	}
}

// updateItemsPerPage updates the items per page for each section.
func (s *RunOverviewSidebar) updateItemsPerPage() {
	for i := range s.sections {
		if s.sections[i].Height > 0 {
			// Height includes title line, so items per page is height - 1.
			s.sections[i].SetItemsPerPage(max(s.sections[i].Height-1, 1))
		} else {
			s.sections[i].SetItemsPerPage(0)
		}
	}
}

// navigateUp moves cursor up within the active section.
func (s *RunOverviewSidebar) navigateUp() {
	if !s.isValidActiveSection() {
		return
	}

	section := &s.sections[s.activeSection]
	section.Up()
}

// navigateDown moves cursor down within the active section.
func (s *RunOverviewSidebar) navigateDown() {
	if !s.isValidActiveSection() {
		return
	}

	section := &s.sections[s.activeSection]
	section.Down()
}

// navigateSection jumps between sections, skipping empty ones.
func (s *RunOverviewSidebar) navigateSection(direction int) {
	if len(s.sections) == 0 {
		return
	}

	prev := s.activeSection
	idx := prev

	// Try each section in the given direction.
	for range len(s.sections) {
		idx += direction
		if idx < 0 {
			idx = len(s.sections) - 1
		} else if idx >= len(s.sections) {
			idx = 0
		}

		// Select first non-empty section.
		if len(s.sections[idx].FilteredItems) > 0 {
			s.setActiveSection(idx)
			return
		}

		// Wrapped around to starting section.
		if idx == prev {
			break
		}
	}

	// No non-empty section found, keep current active.
	s.sections[prev].Active = true
}

// navigatePageUp changes page to previous within active section.
func (s *RunOverviewSidebar) navigatePageUp() {
	if !s.isValidActiveSection() {
		return
	}

	section := &s.sections[s.activeSection]
	section.PageUp()
}

// navigatePageDown changes page to next within active section.
func (s *RunOverviewSidebar) navigatePageDown() {
	if !s.isValidActiveSection() {
		return
	}

	section := &s.sections[s.activeSection]
	section.PageDown()
}

// navigateHome jumps to the first item of the active section.
func (s *RunOverviewSidebar) navigateHome() {
	if !s.isValidActiveSection() {
		return
	}

	section := &s.sections[s.activeSection]
	section.Home()
}

// navigateEnd jumps to the last item of the active section.
func (s *RunOverviewSidebar) navigateEnd() {
	if !s.isValidActiveSection() {
		return
	}

	section := &s.sections[s.activeSection]
	section.End()
}

// selectFirstAvailableItem selects the first item in the first non-empty section.
func (s *RunOverviewSidebar) selectFirstAvailableItem() {
	// Find first non-empty section.
	for i := range s.sections {
		if len(s.sections[i].FilteredItems) > 0 && s.sections[i].ItemsPerPage() > 0 {
			s.setActiveSection(i)
			return
		}
	}

	// No non-empty section found, default to first section.
	s.setActiveSection(0)
}

// restoreSelection attempts to restore the previously selected item.
func (s *RunOverviewSidebar) restoreSelection(previousKey string) {
	// Try to find key-only match in current active section.
	if s.tryRestoreInSection(previousKey) {
		return
	}

	// Could not restore, select first available.
	s.selectFirstAvailableItem()
}

// tryRestoreInSection attempts to restore selection in a specific section.
//
// Returns true if successful.
func (s *RunOverviewSidebar) tryRestoreInSection(key string) bool {
	if s.activeSection < 0 || s.activeSection >= len(s.sections) {
		return false
	}

	section := &s.sections[s.activeSection]
	if section.ItemsPerPage() == 0 {
		return false
	}

	for i, item := range section.FilteredItems {
		keyMatch := item.Key == key

		if keyMatch {
			page := i / section.ItemsPerPage()
			line := i % section.ItemsPerPage()

			section.SetPageAndLine(page, line)
			return true
		}
	}

	return false
}

// deactivateAllSections marks all sections as inactive, removing row highlights.
func (s *RunOverviewSidebar) deactivateAllSections() {
	for i := range s.sections {
		s.sections[i].Active = false
	}
}

// hasActiveSection reports whether any section is currently active.
func (s *RunOverviewSidebar) hasActiveSection() bool {
	for i := range s.sections {
		if s.sections[i].Active {
			return true
		}
	}
	return false
}

// setActiveSection changes the active section and resets navigation state.
func (s *RunOverviewSidebar) setActiveSection(idx int) {
	// Deactivate all sections.
	for i := range s.sections {
		s.sections[i].Active = false
	}

	// Activate target section.
	s.activeSection = idx
	if idx >= 0 && idx < len(s.sections) {
		s.sections[idx].Active = true
	}
}

// isValidActiveSection returns true if the active section index is valid.
func (s *RunOverviewSidebar) isValidActiveSection() bool {
	return s.activeSection >= 0 && s.activeSection < len(s.sections)
}

File: leet/crates/leet-tui/src/run_overview_sidebar_nav.rs
//! Port of `core/internal/leet/runoverviewsidebarnav.go` — section height
//! allocation and navigation for
//! [`RunOverviewSidebar`](crate::run_overview_sidebar::RunOverviewSidebar).

use crate::run_overview_sidebar::RunOverviewSidebar;

// Maximum heights for each section type.
// TODO: dynamically upscale if more space is available.
const SECTION_MAX_HEIGHT_ENVIRONMENT: isize = 12;
const SECTION_MAX_HEIGHT_CONFIG: isize = 20;
const SECTION_MAX_HEIGHT_SUMMARY: isize = 25;

/// Minimum section height when visible (title + 1 item).
const SECTION_MIN_HEIGHT: isize = 2;

impl RunOverviewSidebar {
    /// updateSectionHeights dynamically allocates heights to sections.
    pub(crate) fn update_section_heights(&mut self) {
        if self.height == 0 {
            return;
        }

        let total_available = self.available_height();
        if total_available <= 0 {
            return;
        }

        let desired = self.calculate_desired_heights();
        let total_desired = self.sum_desired_heights(&desired);

        if total_desired > total_available {
            self.scale_heights_proportionally(&desired, total_available);
        } else {
            self.allocate_desired_heights(&desired);
            self.distribute_extra_space(total_available, total_desired);
        }

        self.update_items_per_page();
    }

    /// availableHeight returns the height available for sections.
    fn available_height(&self) -> isize {
        let available_height = self.height - self.header_line_count();

        let active_sections = self.count_active_sections();
        if active_sections == 0 {
            return 0;
        }

        // Account for spacing between sections.
        let mut spacing_between_sections = 0;
        if active_sections > 1 {
            spacing_between_sections = active_sections - 1;
        }

        // Ensure minimum space for all active sections.
        let min_required = active_sections * SECTION_MIN_HEIGHT;
        (available_height - spacing_between_sections).max(min_required)
    }

    /// countActiveSections returns the number of sections with items.
    fn count_active_sections(&self) -> isize {
        let mut count = 0;
        for section in &self.sections {
            if !section.filtered_items.is_empty() {
                count += 1;
            }
        }
        count
    }

    /// calculateDesiredHeights calculates the desired height for each
    /// section.
    fn calculate_desired_heights(&mut self) -> Vec<isize> {
        let max_heights = [
            SECTION_MAX_HEIGHT_ENVIRONMENT,
            SECTION_MAX_HEIGHT_CONFIG,
            SECTION_MAX_HEIGHT_SUMMARY,
        ];

        let mut desired = vec![0isize; self.sections.len()];

        for (i, section) in self.sections.iter_mut().enumerate() {
            let item_count = section.filtered_items.len() as isize;
            if item_count == 0 {
                section.height = 0;
                desired[i] = 0;
                continue;
            }

            // Desired height is item count + 1 (for title), capped at max.
            let max_height = max_heights[i];
            desired[i] = (item_count + 1).min(max_height).max(SECTION_MIN_HEIGHT);
        }

        desired
    }

    /// sumDesiredHeights returns the sum of all desired heights.
    fn sum_desired_heights(&self, desired: &[isize]) -> isize {
        let mut total = 0;
        for &h in desired {
            total += h;
        }
        total
    }

    /// scaleHeightsProportionally scales section heights when total exceeds
    /// available.
    fn scale_heights_proportionally(&mut self, desired: &[isize], total_available: isize) {
        let total_desired = self.sum_desired_heights(desired);
        let scale_factor = total_available as f64 / total_desired as f64;

        let mut allocated = 0;
        for (i, section) in self.sections.iter_mut().enumerate() {
            if desired[i] > 0 {
                // PARITY: Go `int(float64(desired[i]) * scaleFactor)`
                // truncates toward zero; `as isize` matches.
                let mut scaled = (desired[i] as f64 * scale_factor) as isize;
                // Enforce minimum height for visible sections.
                if scaled < SECTION_MIN_HEIGHT && !section.filtered_items.is_empty() {
                    scaled = SECTION_MIN_HEIGHT;
                }
                section.height = scaled;
                allocated += scaled;
            } else {
                section.height = 0;
            }
        }

        // Distribute remainder to last section with items.
        if allocated < total_available {
            let remainder = total_available - allocated;
            self.allocate_remainder(remainder);
        }
    }

    /// allocateDesiredHeights sets each section to its desired height.
    fn allocate_desired_heights(&mut self, desired: &[isize]) {
        for (i, section) in self.sections.iter_mut().enumerate() {
            section.height = desired[i];
        }
    }

    /// distributeExtraSpace distributes unused space to sections that can
    /// use it.
    fn distribute_extra_space(&mut self, total_available: isize, total_desired: isize) {
        let max_heights = [
            SECTION_MAX_HEIGHT_ENVIRONMENT,
            SECTION_MAX_HEIGHT_CONFIG,
            SECTION_MAX_HEIGHT_SUMMARY,
        ];

        let mut extra_space = total_available - total_desired;

        // Try to expand sections from bottom to top (summary, config, env).
        for i in (0..=2usize).rev() {
            if extra_space <= 0 {
                break;
            }
            let section = &mut self.sections[i];
            if section.height == 0 {
                continue;
            }

            let item_count = section.filtered_items.len() as isize;
            let current_items = section.height - 1; // Subtract title line

            // Only expand if we have more items to show.
            if current_items < item_count {
                let max_increase =
                    (max_heights[i] - section.height).min(item_count + 1 - section.height);
                let increase = max_increase.min(extra_space);

                section.height += increase;
                extra_space -= increase;
            }
        }
    }

    /// allocateRemainder distributes remaining space to the last section
    /// with items.
    fn allocate_remainder(&mut self, remainder: isize) {
        // Try sections from bottom to top.
        for i in (0..=2usize).rev() {
            if !self.sections[i].filtered_items.is_empty() && self.sections[i].height > 0 {
                self.sections[i].height += remainder;
                return;
            }
        }
    }

    /// updateItemsPerPage updates the items per page for each section.
    fn update_items_per_page(&mut self) {
        for section in &mut self.sections {
            if section.height > 0 {
                // Height includes title line, so items per page is height - 1.
                section.set_items_per_page((section.height - 1).max(1));
            } else {
                section.set_items_per_page(0);
            }
        }
    }

    /// navigateUp moves cursor up within the active section.
    pub(crate) fn navigate_up(&mut self) {
        if !self.is_valid_active_section() {
            return;
        }

        let section = &mut self.sections[self.active_section as usize];
        section.up();
    }

    /// navigateDown moves cursor down within the active section.
    pub(crate) fn navigate_down(&mut self) {
        if !self.is_valid_active_section() {
            return;
        }

        let section = &mut self.sections[self.active_section as usize];
        section.down();
    }

    /// navigateSection jumps between sections, skipping empty ones.
    pub(crate) fn navigate_section(&mut self, direction: isize) {
        if self.sections.is_empty() {
            return;
        }

        let prev = self.active_section;
        let mut idx = prev;

        // Try each section in the given direction.
        for _ in 0..self.sections.len() {
            idx += direction;
            if idx < 0 {
                idx = self.sections.len() as isize - 1;
            } else if idx >= self.sections.len() as isize {
                idx = 0;
            }

            // Select first non-empty section.
            if !self.sections[idx as usize].filtered_items.is_empty() {
                self.set_active_section(idx);
                return;
            }

            // Wrapped around to starting section.
            if idx == prev {
                break;
            }
        }

        // No non-empty section found, keep current active.
        self.sections[prev as usize].active = true;
    }

    /// navigatePageUp changes page to previous within active section.
    pub(crate) fn navigate_page_up(&mut self) {
        if !self.is_valid_active_section() {
            return;
        }

        let section = &mut self.sections[self.active_section as usize];
        section.page_up();
    }

    /// navigatePageDown changes page to next within active section.
    pub(crate) fn navigate_page_down(&mut self) {
        if !self.is_valid_active_section() {
            return;
        }

        let section = &mut self.sections[self.active_section as usize];
        section.page_down();
    }

    /// navigateHome jumps to the first item of the active section.
    // PHASE-5: called from the run/workspace key handlers
    // (runhandlers.go:466, workspacehandlers.go:835).
    #[allow(dead_code)]
    pub(crate) fn navigate_home(&mut self) {
        if !self.is_valid_active_section() {
            return;
        }

        let section = &mut self.sections[self.active_section as usize];
        section.home();
    }

    /// navigateEnd jumps to the last item of the active section.
    // PHASE-5: called from the run/workspace key handlers
    // (runhandlers.go:482, workspacehandlers.go:855).
    #[allow(dead_code)]
    pub(crate) fn navigate_end(&mut self) {
        if !self.is_valid_active_section() {
            return;
        }

        let section = &mut self.sections[self.active_section as usize];
        section.end();
    }

    /// selectFirstAvailableItem selects the first item in the first
    /// non-empty section.
    pub(crate) fn select_first_available_item(&mut self) {
        // Find first non-empty section.
        for i in 0..self.sections.len() {
            if !self.sections[i].filtered_items.is_empty() && self.sections[i].items_per_page() > 0
            {
                self.set_active_section(i as isize);
                return;
            }
        }

        // No non-empty section found, default to first section.
        self.set_active_section(0);
    }

    /// restoreSelection attempts to restore the previously selected item.
    pub(crate) fn restore_selection(&mut self, previous_key: &str) {
        // Try to find key-only match in current active section.
        if self.try_restore_in_section(previous_key) {
            return;
        }

        // Could not restore, select first available.
        self.select_first_available_item();
    }

    /// tryRestoreInSection attempts to restore selection in a specific
    /// section.
    ///
    /// Returns true if successful.
    fn try_restore_in_section(&mut self, key: &str) -> bool {
        if self.active_section < 0 || self.active_section as usize >= self.sections.len() {
            return false;
        }

        let section = &mut self.sections[self.active_section as usize];
        if section.items_per_page() == 0 {
            return false;
        }

        // PARITY: Go returns from inside the range loop on the first key
        // match; the index is found first here to satisfy the borrow
        // checker (`SetPageAndLine` needs `&mut`).
        let mut found: Option<isize> = None;
        for (i, item) in section.filtered_items.iter().enumerate() {
            let key_match = item.key == key;

            if key_match {
                found = Some(i as isize);
                break;
            }
        }

        if let Some(i) = found {
            let page = i / section.items_per_page();
            let line = i % section.items_per_page();

            section.set_page_and_line(page, line);
            return true;
        }

        false
    }

    /// deactivateAllSections marks all sections as inactive, removing row
    /// highlights.
    pub(crate) fn deactivate_all_sections(&mut self) {
        for section in &mut self.sections {
            section.active = false;
        }
    }

    /// hasActiveSection reports whether any section is currently active.
    pub(crate) fn has_active_section(&self) -> bool {
        for section in &self.sections {
            if section.active {
                return true;
            }
        }
        false
    }

    /// setActiveSection changes the active section and resets navigation
    /// state.
    pub(crate) fn set_active_section(&mut self, idx: isize) {
        // Deactivate all sections.
        for section in &mut self.sections {
            section.active = false;
        }

        // Activate target section.
        self.active_section = idx;
        if idx >= 0 && (idx as usize) < self.sections.len() {
            self.sections[idx as usize].active = true;
        }
    }

    /// isValidActiveSection returns true if the active section index is
    /// valid.
    pub(crate) fn is_valid_active_section(&self) -> bool {
        self.active_section >= 0 && (self.active_section as usize) < self.sections.len()
    }
}
